view_conversations/conv-viewer-webapi/src/db_types.ts

```diff
@@ -0,0 +1,172 @@
+}
+
+
+
+class IntFieldHandler extends SpannerFieldHandler<number> {
+  public fromSpannerResultField(field:spanner.ResultField) : number | null {
+    if(typeof(field) === 'string') {
+      throw Error(`For ${this.fieldName}: expected field.value:string, but field:string`);
+    } else if (field instanceof Date) {
+      throw Error(`For ${this.fieldName}: expected field.value:string, but field:Date`);
+    } else if (field === null) {
+      return null;
+    }
+    return parseInt((field as spanner.ResultValue).value, 10);
+  }
+}
+
+class FloatFieldHandler extends SpannerFieldHandler<number> {
+  public fromSpannerResultField(field:spanner.ResultField) : number | null {
+    if(typeof(field) === 'string') {
+      throw Error(`For ${this.fieldName}: expected field.value:string, but field:string`);
+    } else if (field instanceof Date) {
+      throw Error(`For ${this.fieldName}: expected field.value:string, but field:Date`);
+    } else if (field === null) {
+      return null;
+    }
+    return parseFloat((field as spanner.ResultValue).value);
+  }
+}
+
+class TimestampFieldHandler extends SpannerFieldHandler<Date> {
+  public fromSpannerResultField(field:spanner.ResultField) : Date | null {
+    if (!(field === null || field instanceof Date)) {
+      throw Error(`For ${this.fieldName}: expected field:Date, but field:${typeof(field)}`);
+    }
+    return field;
+  }
+}
+
+interface HandlerSet { [fieldName:string] : SpannerFieldHandler<any>; };
+interface ParsedOutput { [fieldName:string] : {}; };
+
+const handlers : Array<SpannerFieldHandler<{}>> = [
+  new StringFieldHandler('id'),
+  new StringFieldHandler('ancestor_id'),
+  new ArrayFieldHandler('authors'),
+  new BytesFieldHandler('cleaned_content'),
+  new BytesFieldHandler('content'),
+  new StringFieldHandler('conversation_id'),
+  new IntFieldHandler('indentation'),
+  new IntFieldHandler('page_id'),
+  new StringFieldHandler('page_title'),
+  new StringFieldHandler('parent_id'),
+  new StringFieldHandler('replyTo_id'),
+  new IntFieldHandler('rev_id'),
+  new TimestampFieldHandler('timestamp'),
+  new StringFieldHandler('type'),
+  new IntFieldHandler('user_id'),
+  new StringFieldHandler('user_text'),
+];
+
+function addHandler(inputHandlers : HandlerSet, handler : SpannerFieldHandler<{}>)
+    : HandlerSet {
+  inputHandlers[handler.fieldName] = handler;
+  return inputHandlers;
+}
+const handlerSet = handlers.reduce<HandlerSet>(addHandler, {});
+
+export function parseOutputRows<T>(rows: spanner.ResultRow[]) : T {
+  const output : ParsedOutput[] = []
+  for (const row of rows) {
+    const ret: { [fieldName:string] : {} } = {};
+    for (const field of row) {
+      if(!(field.name in handlerSet)) {
+        console.error(`Field ${field.name} does not have a handler and so cannot be interpreted.`);
+        break;
+      }
+      ret[field.name] = handlerSet[field.name].fromSpannerResultField(field.value);
+    }
+    output.push(ret)
+  }
+  return output as any as T;
```

Comment:
Generics really shouldn't be used in this way, and this type of casting is unsafe. It looks like you are always using OutputRow as the generic type anyway in all the places you call this, so why not just use that as your type here and declare const output: OutputRow[] at the start? That way this can be a type safe function. I think that would also eliminate the need to check the names of the fields.

If you do this you might run into an issue where the compiler complains that not all fields are present since you are building the OutputRow field by field. There are a few options to workaround this. One, you can define the type OutputRow such that fields are optional (this is done with the syntax :? when defining each field) or you can use a partial type, e.g. const ret: Partial<OutputRow> = {}; (More on that here: https://netbasal.com/getting-to-know-the-partial-type-in-typescript-ecfcfbc87cb6)

Reply:
I think another problem here is the compiler complains about output not being the template type T when changing output type to OutputRow[].

Reply:
Right, if you did this you would have to remove the generics.